src/repository/opamDownload.ml

```diff
@@ -188,8 +188,18 @@ let download_as ?quiet ?validate ~overwrite ?compress ?checksum url dst =
       ()
 
 let download ?quiet ?validate ~overwrite ?compress ?checksum url dstdir =
+  let base =
+    let base = OpamUrl.basename url in
+    if Sys.win32 then
+      let f c =
+        if OpamStd.Sys.is_valid_basename_char c then c else '_'
+      in
+      String.map f base
+    else
+      base
+  in
   let dst =
-    OpamFilename.(create dstdir (Base.of_string (OpamUrl.basename url)))
+    OpamFilename.(create dstdir (Base.of_string base))
```

Comment:
I'm wondering if we should have that check/change in `OpamFilename.Base` directly, like that we are sure that every file created is valid. But can we have that change (forbidden char -> `_`) for all files?

Reply:
I don't think we can ever automatically change it, but we may want to do things to cause better error messages. For example, if an opam file actually specifies `extra-source "zstd-detection?full_index=1.patch" {` then I don't think opam should ever do anything to make this work, but it perhaps fail with a better error message if this is encountered.

Reply:
(the key point here is that this filename is internal to opam)